common/djangoapps/entitlements/models.py

```diff
@@ -302,6 +304,45 @@ def get_active_entitlements_for_user(cls, user):
             enrollment_course_run=None
         ).select_related('user').select_related('enrollment_course_run')
 
+    @classmethod
+    def check_for_existing_entitlement_and_enroll(cls, user, course_run_key):
```

Comment:
I think there's some clean-code 'functions should only do one thing' that could be applied here.
It could be really useful to break this into two independent functions:
```python
def is_user_entitled_to_course(user, course_run_key):
```
and
```python
def enroll_entitled_user_in_course(user, course_run_key):
```

this would require a tweak of common/djangoapps/student/views/management.py and wherever else we've called it from 
```python
if CourseEntitlement.check_for_existing_entitlement_and_enroll(user=user, course_run_key=course_id):
  return HttpResponse(reverse('courseware', args=[unicode(course_id)]))
```
to
```python
if CourseEntitlement.is_user_entitled_to_course(user=user, course_run_key=course_id):
  CourseEntitlement.enroll_entitled_user_in_course(user, course_run_key)
  return HttpResponse(reverse('courseware', args=[unicode(course_id)]))
```
I know this ends up being a little more verbose in the views, but I think its more dynamic and would allow for cleaner handling of the case where the user isn't entitled. 
Also we'd get two functions that are useful in more cases than just right here.

Reply:
If I split this into two methods I would end up wrapping when into a single method for the two places it is used. 

I do see where these methods could be of use in the future.  Let me take a look at splitting it up a bit.

Reply:
Oooh wrapping it is a good idea, best of both words, short syntax in views and reproducable functions in CourseEntitlement

Reply:
Take a look at my rework.  The one issue with splitting this into two methods in this way is that I would need to retrieve the entitlement twice.